### PR TITLE
Adding support for parallel tools based on the 3rd party library mpiFileUtils (libMFU)

### PR DESCRIPTION
**Pull request by rawarren**
This is a first draft for adding a new parallel tool (dh5walk) which is based on other tools found in the mpifileutils project (see: https://github.com/hpc/mpifileutils.git).   The new 'dh5walk' tool is intended to run as an MPI based application and provides a parallel wrapper to allow individual MPI ranks to launch any of the unmodified serial HDF5 tools on hdf5 files found in a shared directory.   The effect is that by assigning a tool instance to each mpi rank for each independent file, that parallelism can provide speedups over simple scripting mechanisms.

### CODE REVIEW

File: MANIFEST [hunk 1]
```diff
@@ -3588,6 +3594,10 @@
 
 ./config/cmake_ext_mod/ConfigureChecks.cmake
 ./config/cmake_ext_mod/CTestCustom.cmake
+./config/cmake_ext_mod/.cmake
```
Comment (byrnHDF) [inline MANIFEST]: ".cmake"? Missing filename?

File: MANIFEST [hunk 2]
```diff
@@ -3588,6 +3594,10 @@
...
+./config/cmake_ext_mod/FindCIRCLE.cmake
```
Comment (byrnHDF) [inline MANIFEST]: Are these new files useful for other hdf products (hdf4, h4h5tools, ...)? If only for hdf5, then they should be put into the ./config/cmake directory.

File: tools/src/dh5walk/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+  set (H5_DEP_EXECUTABLES dh5walk)
+endif ()
+
+if (BUILD_SHARED_LIBS)
+  add_executable (dh5walk-shared ${HDF5_TOOLS_SRC_DH5WALK_SOURCE_DIR}/dh5walk.c)
+  target_include_directories (dh5walk-shared PRIVATE "${HDF5_TOOLS_DIR}/lib;${HDF5_SRC_DIR};${HDF5_SRC_BINARY_DIR};${CIRCLE_INCLUDE_DIR};$<$<BOOL:${HDF5_ENABLE_PARALLEL}>:${MPI_C_INCLUDE_DIRS}>")
+  target_compile_options(dh5walk-shared PRIVATE "${HDF5_CMAKE_C_FLAGS}")
+  TARGET_C_PROPERTIES (dh5walk-shared SHARED)
+  target_link_libraries (dh5walk-shared PRIVATE ${HDF5_TOOLS_LIBSH_TARGET} ${HDF5_LIBSH_TARGET} m ${MFU_LIBRARY} "$<$<BOOL:${HDF5_ENABLE_PARALLEL}>:${MPI_C_LIBRARIES}>")
```
Comment (byrnHDF) [inline tools/src/dh5walk/CMakeLists.txt]: Is that "m" supposed to be there between ${HDF5_LIBSH_TARGET} and ${MFU_LIBRARY}?

File: CMakeLists.txt [hunk 4]
```diff
@@ -1017,6 +1017,27 @@ if (EXISTS "${HDF5_SOURCE_DIR}/tools" AND IS_DIRECTORY "${HDF5_SOURCE_DIR}/tools
   if (HDF5_BUILD_TOOLS)
     add_subdirectory (tools)
   endif ()
+  option (HDF5_BUILD_PARALLEL_TOOLS  "Build Parallel HDF5 Tools" OFF)
```
Comment (byrnHDF) [inline CMakeLists.txt]: I think HDF5_BUILD_TOOLS should be a prerequisite of HDF5_BUILD_PARALLEL_TOOLS, maybe?
Does it make sense to build this tool without building the regular tools?

**Comment by lrknox** [comment]
Tests failed because 2 files in the MANIFEST aren't found in the github source:  dh5walk and cmake_ext_mod/.cmake.  The MANIFEST lists which source files should be included in a source tarball.  It can't include generated files that aren't committed to git.

File: tools/src/dh5walk/dh5walk.1 [hunk 5]
```diff
@@ -0,0 +1,42 @@
+.TH "dh5walk" 1
+.SH NAME
+dh5walk \- Provides a means of extending HDF5 tools by using parallelism on groups of files.
+.SH SYNOPSIS
+dh5walk [OPTIONS] -T h5tool [H5TOOL_options...] 
+.SH DESCRIPTION
+dh5walk utilizes the mpiFileUtils library to invoke a selected HDF5 tool on a collection of files.  The mpiFileUtils library provides the facilities to walk directory trees and provide a selection of files contained therein.  This selection can be filtered in various ways.  At present, dh5walk filters the original file selection to include include only HDF5 formatted files.  The resulting collection or collections can be utilized as the file inputs to the selected h5tool.
```
Comment (lrknox) [inline tools/src/dh5walk/dh5walk.1]: redundant include include

**Review by byrnHDF — CHANGES_REQUESTED** [review]

**Comment by epourmal** [comment]
@rawarren Maybe I missed it...
We need to check in User's Guide in both docx and PDF formats to the h5docs repo https://github.com/HDFGroup/hdf5doc/tree/master/RFCs/HDF5/tools(just create a new directory. Thank you!

File: CMakeLists.txt [hunk 6]
```diff
@@ -1016,6 +1016,27 @@ if (EXISTS "${HDF5_SOURCE_DIR}/tools" AND IS_DIRECTORY "${HDF5_SOURCE_DIR}/tools
...
   if (HDF5_BUILD_TOOLS)
     add_subdirectory (tools)
+    option (HDF5_BUILD_PARALLEL_TOOLS  "Build Parallel HDF5 Tools" OFF)
+    if (HDF5_BUILD_PARALLEL_TOOLS AND HDF5_ENABLE_PARALLEL)
+      find_package(MFU REQUIRED)
+      if (MFU_FOUND)
+        set (H5_HAVE_LIBMFU 1)
+        set (H5_HAVE_MFU_H 1)
+        set (CMAKE_REQUIRED_INCLUDES "${MFU_INCLUDE_DIR}")
+	  endif ()
```
Comment (byrnHDF) [inline CMakeLists.txt:1026]: indentation - minor issue.

File: tools/src/dh5walk/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+project (HDF5_TOOLS_SRC_DH5WALK C)
+
+# --------------------------------------------------------------------
+# Add the dh5walk and test executables
+# --------------------------------------------------------------------
+if (NOT ONLY_SHARED_LIBS)
+  add_executable (dh5walk ${HDF5_TOOLS_SRC_DH5WALK_SOURCE_DIR}/dh5walk.c)
+  add_custom_command(
+	OUTPUT ${HDF5_TOOLS_DIR}/test/demo_destfiles.test
+	COMMAND bash -c ${HDF5_TOOLS_SRC_DH5WALK_SOURCE_DIR}/copy_demo_files.sh
```
Comment (byrnHDF) [inline tools/src/dh5walk/CMakeLists.txt:11]: Can we convert this script to use the cmake copy method? Search for "HDFTEST_COPY_FILE" in cmaketests.cmake files. This will not execute on windows, and would need to be guarded against (which we do for the TEST_SHELL_SCRIPTS)

**Comment by byrnHDF** [comment]
I'm confused! Why are we reinventing the usage/help and cmdline arg processing? Can't we use the library functions? Also, we are
limiting it to *nix usage only, we should instead implement it generally and then restrict it to usable platforms as necessary.

**Review by byrnHDF — CHANGES_REQUESTED** [review]
I would be glad to help with the CMake coding.

**Comment by byrnHDF** [comment]
In addition, where are the tests? What is the working directory?

File: tools/src/dh5walk/Makefile.am [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+	$(SHELL) $(abs_builddir)/copy_demo_files.sh $(abs_top_srcdir)/tools/test $(abs_top_srcdir)/tools/src/h5diff $(abs_top_srcdir)/tools/src/h5ls;
+
+# Include src directory
+AM_CPPFLAGS+=-I$(top_srcdir)/src -I$(top_srcdir)/tools/lib
+
+# These are our main targets, the tools
+bin_PROGRAMS=dh5walk
+bin_SCRIPTS=install-examples
+
+# Add h5stat specific linker flags here
```
Comment (byrnHDF) [inline tools/src/dh5walk/Makefile.am:32]: comment tool name?

File: tools/src/dh5walk/dh5walk.c [hunk 9]
```diff
@@ -0,0 +1,1685 @@
...
+#include "libcircle.h"
+#include "dtcmp.h"
+#include "mfu.h"
+#include "mfu_flist.h"
+#include "mfu_errors.h"
+#include "mfu_flist_internal.h"
+
+typedef int64_t hid_t;
+#define H5P_DEFAULT (hid_t)0
+extern int H5F__is_hdf5(const char *name, hid_t fapl);
```
Comment (byrnHDF) [inline tools/src/dh5walk/dh5walk.c]: extern should not be used - windows will not link correctly, use H5DLL decoration.

File: tools/src/dh5walk/dh5walk.c [hunk 10]
```diff
@@ -0,0 +1,1685 @@
+#include <dirent.h>
```
Comment (byrnHDF) [inline tools/src/dh5walk/dh5walk.c]: where is the hdf5 header and because this is in tools use the h5tools include

File: tools/src/dh5walk/dh5walk.c [hunk 11]
```diff
@@ -0,0 +1,1685 @@
...
+#include <limits.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <string.h>
+#include <unistd.h>
+#include <getopt.h>
```
Comment (byrnHDF) [inline tools/src/dh5walk/dh5walk.c]: Use the tools library for getopt - see other tools.

File: tools/src/dh5walk/dh5walk.c [hunk 12]
```diff
@@ -0,0 +1,1685 @@
...
+
+#include "libcircle.h"
+#include "dtcmp.h"
+#include "mfu.h"
+#include "mfu_flist.h"
+#include "mfu_errors.h"
+#include "mfu_flist_internal.h"
+
+typedef int64_t hid_t;
+#define H5P_DEFAULT (hid_t)0
```
Comment (byrnHDF) [inline tools/src/dh5walk/dh5walk.c]: This define is not needed if you include h5tools (which includes hdf5). Do not reinvent the library.

File: tools/src/dh5walk/dh5walk.c [hunk 13]
```diff
@@ -0,0 +1,1685 @@
...
+#define MAX_DISTRIBUTE_SEPARATORS 128
+struct distribute_option {
+    int      separator_number;
+    uint64_t separators[MAX_DISTRIBUTE_SEPARATORS];
+};
+
+static void
+save_command(char *argv0)
+{
+    assert(argv0);
```
Comment (byrnHDF) [inline tools/src/dh5walk/dh5walk.c:99]: using hdf includes allow use of the HD prefix to lessen the cross-platform issues.

File: tools/src/dh5walk/dh5walk.c [hunk 14]
```diff
@@ -0,0 +1,1685 @@
...
+    printf("\n");
+    printf("Usage: %s [options] <path> ...\n", user_cmd);
+#ifdef DAOS_SUPPORT
+    printf("\n");
+    printf("DAOS paths can be specified as:\n");
+    printf("       daos://<pool>/<cont>[/<path>] | <UNS path>\n");
+#endif
+    printf("\n");
+    printf("Options:\n");
+    printf("  -i, --input <file>      - read list from file\n");
```
Comment (byrnHDF) [inline tools/src/dh5walk/dh5walk.c]: We don't publish the short options anymore in our tools.

File: tools/src/dh5walk/dh5walk.c [hunk 15]
```diff
@@ -0,0 +1,1685 @@
...
+    /* return pointer to element if index is within range */
+    uint64_t max = flist->list_count;
+    if (idx < max) {
+        elem_t *elem = flist->list_index[idx];
+        return elem;
+    }
+    return NULL;
+}
+
+#ifdef VERBOSE
```
Comment (byrnHDF) [inline tools/src/dh5walk/dh5walk.c:406]: The tools library has "debug" macros that can be used for debugging. OR should this be a command line option?

File: tools/src/dh5walk/dh5walk.c [hunk 16]
```diff
@@ -0,0 +1,1685 @@
...
+    char  filepath[1024];
+    char *toolname   = argv[0];
+    char *buf        = NULL;
+    int   use_stdout = 0;
+
+    /* create a copy of the 1st file passed to the application */
+    strcpy(filepath, fname);
+
+#if 0
+    if (!log_output_in_single_file && !log_stdout_in_file)
```
Comment (byrnHDF) [inline tools/src/dh5walk/dh5walk.c:935]: tools library has options for redirecting to files or stdout/stderr

File: tools/src/dh5walk/dh5walk.c [hunk 17]
```diff
@@ -0,0 +1,1685 @@
...
+             */
+            log_len = strlen(logpath);
+            strcpy(logErrors, logpath);
+            strcpy(&logErrors[log_len - 3], "err");
+        }
+        if (mfu_debug_level == MFU_LOG_VERBOSE) {
+            printf("\tCreating logfile: %s\n", logpath);
+            fflush(stdout);
+        }
+        pid = fork();
```
Comment (byrnHDF) [inline tools/src/dh5walk/dh5walk.c:1118]: There is no "fork" on windows. if this is the only way, then we need to use the technique used in some tests (see the swmr tests) to gracefully compile on windows. Otherwise, avoid using fork. Why is this needed?

File: tools/src/dh5walk/dh5walk.c [hunk 18]
```diff
@@ -0,0 +1,1685 @@
...
+        int         fileindex[256];
+        int         filecount = 0;
+        int         token     = 0;
+        struct stat statbuf;
+
+        char *eol = strchr(linebuf, '\n');
+        if (eol) {
+            *eol = '\0';
+        }
+        cmdline = strdup(linebuf);
```
Comment (byrnHDF) [inline tools/src/dh5walk/dh5walk.c]: Is there a matching free for cmdline?

**Review by byrnHDF — CHANGES_REQUESTED** [review]

File: tools/src/dh5walk/dh5walk.c [hunk 19]
```diff
@@ -0,0 +1,1685 @@
...
+
+show_help:
+    /* print usage if we need to */
+    if (usage) {
+        if (rank == 0) {
+            print_usage();
+        }
+        mfu_finalize();
+        MPI_Finalize();
+        return 1;
```
Comment (byrnHDF) [inline tools/src/dh5walk/dh5walk.c]: Why return 1 - asking to show help should not be an error, only when there is an actual command error should a 1 be returned. Use our other tools as templates.

File: tools/src/dh5walk/dh5walk.c [hunk 20]
```diff
@@ -0,0 +1,1685 @@
...
+    /* filter files to only include hdf5 files */
+    if (flist1) {
+        filter_hdf_files(&flist1, NULL, 0, 0);
+    }
+    if (flist2) {
+        filter_hdf_files(&flist2, NULL, 0, 0);
+    }
+
+    /* if (numpaths > 1)
+     * In a case where we requeire the list indices of files from multiple
```
Comment (byrnHDF) [inline tools/src/dh5walk/dh5walk.c:1586]: SP

File: tools/src/dh5walk/copy_demo_files.sh [hunk 21]
```diff
@@ -0,0 +1,421 @@
+#! /bin/sh
```
Comment (byrnHDF) [inline tools/src/dh5walk/copy_demo_files.sh:1]: this is just a a test script and should be in the tools/test/dh5walk folder.

File: tools/src/dh5walk/dh5walk.c [hunk 22]
```diff
@@ -1,25 +1,20 @@
...
-#include <pwd.h> /* for getpwent */
-#include <grp.h> /* for getgrent */
-#include <errno.h>
-#include <assert.h>
-#include <inttypes.h>
-#include <string.h>
-#include <math.h>
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
+ * Copyright by The HDF Group.                                               *
+ * Copyright by the Board of Trustees of the University of Illinois.         *
```
Comment (lrknox) [inline tools/src/dh5walk/dh5walk.c]: The "Copyright for the Board of Trustees of the University of Illinois" line shouldn't be added for files created after 2006.

File: tools/test/CMakeLists.txt [hunk 23]
```diff
@@ -1,6 +1,11 @@
 cmake_minimum_required (VERSION 3.12)
 project (HDF5_TOOLS_TEST C)
 
+#-- Add the dh5walk tests
+if (HDF5_BUILD_PARALLEL_TOOLS AND HDF5_ENABLE_PARALLEL)
```
Comment (byrnHDF) [inline tools/test/CMakeLists.txt:5]: Just clarification; this tool only works with parallel library?

File: tools/test/dh5walk/Makefile [hunk 24]
```diff
@@ -0,0 +1,1410 @@
+# Makefile.in generated by automake 1.16.1 from Makefile.am.
```
Comment (byrnHDF) [inline tools/test/dh5walk/Makefile:1]: Makefile should not be committed.

**Comment by lrknox** [comment]
Continued and merged to develop in PR #1177
